base/atomics.jl

```diff
@@ -344,29 +353,29 @@ for typ in atomictypes
     ilt = llvmtypes[inttype(typ)]
     rt = "$lt, $lt*"
     irt = "$ilt, $ilt*"
-    @eval getindex(x::Atomic{$typ}) =
+    @eval getindex(x::Atomic{V}) where {V <: $typ} =
         llvmcall($"""
                  %ptr = inttoptr i$WORD_SIZE %0 to $lt*
                  %rv = load atomic $rt %ptr acquire, align $(gc_alignment(typ))
                  ret $lt %rv
-                 """, $typ, Tuple{Ptr{$typ}}, unsafe_convert(Ptr{$typ}, x))
-    @eval setindex!(x::Atomic{$typ}, v::$typ) =
+                 """, V, Tuple{Ptr{V}}, unsafe_convert(Ptr{V}, x))
+    @eval setindex!(x::Atomic{V}, v::V) where {V <: $typ} =
         llvmcall($"""
                  %ptr = inttoptr i$WORD_SIZE %0 to $lt*
                  store atomic $lt %1, $lt* %ptr release, align $(gc_alignment(typ))
                  ret void
-                 """, Cvoid, Tuple{Ptr{$typ}, $typ}, unsafe_convert(Ptr{$typ}, x), v)
+                 """, Cvoid, Tuple{Ptr{V}, V}, unsafe_convert(Ptr{V}, x), v)
 
     # Note: atomic_cas! succeeded (i.e. it stored "new") if and only if the result is "cmp"
-    if typ <: Integer
-        @eval atomic_cas!(x::Atomic{$typ}, cmp::$typ, new::$typ) =
+    if typ <: Union{Integer, Ptr}
```

Comment:
This should probably be `typ <: Integer || typ <: Ptr`. Otherwise it includes the type `Union{Integer, Ptr}`.